src/main/java/com/bookstore/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+package com.bookstore.service;
+
+import com.bookstore.dto.CartItemRequestDto;
+import com.bookstore.dto.ShoppingCartDto;
+import com.bookstore.dto.ShoppingCartRequestDto;
+import com.bookstore.entity.Book;
+import com.bookstore.entity.CartItemKey;
+import com.bookstore.entity.ShoppingCart;
+import com.bookstore.entity.User;
+import com.bookstore.exception.EntityNotFoundException;
+import com.bookstore.mapper.ShoppingCartMapper;
+import com.bookstore.repository.CartItemRepository;
+import com.bookstore.repository.ShoppingCartRepository;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final LocaleService localeService;
+    private final BookService bookService;
+
+    @Override
+    public void createFor(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCart.setId(user.getId());
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto save(ShoppingCartRequestDto dto, User user) {
+        List<Long> newBookIds = dto.getCartItems().stream()
+                .map(CartItemRequestDto::getBookId)
+                .toList();
+        bookService.validateBooksExistence(newBookIds);
+
+        ShoppingCart shoppingCart = shoppingCartMapper.toModel(dto);
+        shoppingCart.setUser(user);
+        shoppingCart.setId(user.getId());
+
+        shoppingCart.setCartItems(shoppingCart.getCartItems().stream()
```

Comment:
check if shop cart already has the book, if yes - increase quantity, otherwise - add to cart